resource/compute_dynamic_params.py

```diff
@@ -0,0 +1,182 @@
+                choices += [connection_choice(connection)]
+            choices += [indent(dataset_choice(dataset)) for dataset in datasets_per_connection[connection]]
+        return {'choices': choices}
+
+    if parameter_name == 'stage':
+        choices = []
+        for connection in datasets_per_connection:
+            if multiple_connections:
+                choices += [connection_choice(connection)]
+            try:
+                choices += [indent(stage_choice(row)) for row in get_stages(connection=connection)]
+            except Exception as e:
+                logging.exception('Error while fetching Snowflake stages on connection `%s`', connection, exc_info=e)
+                choices += [indent(failed_connection("stages"))]
+        return {"choices": choices}
+
+    if parameter_name == 'file_format':
+        choices = [indent(default_format_choice)]
+        for connection in datasets_per_connection:
+            if multiple_connections:
+                choices += [connection_choice(connection)]
+            try:
+                choices += [indent(file_format_choice(row)) for row in get_file_formats(connection=connection)]
+            except Exception as e:
+                logging.exception('Error while fetching Snowflake file formats on connection `%s`', connection, exc_info=e)
+                choices += [indent(failed_connection("file formats"))]
+        return {"choices": choices}
+
+
+def macro_from_dataset_params(parameter_name, dataset_name):
+    if parameter_name == 'dataset':  # this param is only used to display the dataset name to the user in the macro modal
+        return {"choices": [{"value": "default", "label": dataset_name}]}
+
+    if parameter_name == 'stage':
+        if not is_dataset_valid(dataset_name):
+            return {"choices": [invalid_dataset_choice]}
+
+        try:
+            choices = [stage_choice(row) for row in get_stages(dataset=dataset_name)]
+        except Exception as e:
+            logging.exception('Error while fetching Snowflake stages for dataset `%s`', dataset_name, exc_info=e)
+            choices = [failed_connection("stages")]
+        return {"choices": choices}
+
+    if parameter_name == 'file_format':
+        if not is_dataset_valid(dataset_name):
+            return {"choices": [invalid_dataset_choice]}
+
+        try:
+            choices = [default_format_choice] + \
+                      [file_format_choice(row) for row in get_file_formats(dataset=dataset_name)]
+        except Exception as e:
+            logging.exception('Error while fetching Snowflake file formats for dataset `%s`', dataset_name, exc_info=e)
+            choices = [failed_connection("file formats")]
+        return {"choices": choices}
+
+
+def get_snowflake_datasets():
+    project_key = default_project_key()
+    project = api_client().get_project(project_key)
+    return filter(lambda dataset: dataset.type == 'Snowflake', project.list_datasets())
```

Comment:
In Python, we prefer the structure as the structure you use returns a `filter object`
```suggestion
    return [ds for ds in project.list_datasets() if ds.type == 'Snowflake']
```